src/js/gpio.js

```diff
@@ -47,47 +51,140 @@ GPIO.release = function() {
-    });
+function getModeCode(mode) {
+  if (mode == 'pullup') return gpioctl.PULLUP;
+  if (mode == 'pulldn') return gpioctl.PULLDN;
+  if (mode == 'float') return gpioctl.FLOAT;
+  if (mode == 'pushpull') return gpioctl.PUSHPULL;
+  if (mode == 'opendrain') return gpioctl.OPENDRAIN;
+  if (mode == undefined || mode == 'none' || mode == '') return gpioctl.NONE;
+  return -1;
+}
+
+
+GPIO.setPin = function(pinNumber, direction, mode, callback) {
+  if (util.isFunction(mode)) {
+    callback = mode;
+    mode = undefined;
+  }
+  var err, result = 0;
+  var dircode = getDirectionCode(direction);
+  var modecode = getModeCode(mode);
+
+  if (dircode < 0 || modecode < 0) {
+    result = gpioctl.ERR_INVALIDPARAM;
   }
   else {
-    return err;
+    result = gpioctl.setPin(pinNumber, dircode, modecode, callback);
+  }
+  if (result < 0) {
+    err = new GpioError(result, gpioctl.errMessage(result));
+    if (util.isFunction(callback)) {
+      callback(err);
+    }
+    else {
+      throw err;
+    }
   }
 };
 
 
-GPIO.write = function(portpin, val, callback) {
-  var err = gpioctl.writepin(portpin, val);
-
-  if (util.isFunction(callback)) {
-    process.nextTick(function() {
+GPIO.writePin = function(pinNumber, value, callback) {
+  var result = gpioctl.writepin(pinNumber, value, callback);
+  var err;
+  if (result < 0) {
+    err = new GpioError(result, gpioctl.errMessage(result));
+    if (util.isFunction(callback)) {
       callback(err);
-    });
+    }
+    else {
+      throw err;
+    }
   }
-  else {
-    return err;
+};
+
+
+GPIO.readPin = function(pinNumber, callback) {
+  var result = 0;
+  if (!util.isFunction(callback)) {
+    result = gpioctl.ERR_INVALIDPARAM;
+    throw new GpioError(result, gpioctl.errMessage(result));
+  }
+  result = gpioctl.readPin(pinNumber, callback);
+  if (result < 0) {
+    callback(err, false);
   }
 };
 
 
-GPIO.read = function(portpin, callback) {
-  var err = gpioctl.readpin(portpin);
-  var value = err;
+GPIO.setPort = function(portNumber, direction, mode, callback) {
+  if (util.isFunction(mode)) {
+    callback = mode;
+    mode = undefined;
+  }
+  var err, result = 0;
+  var dircode = getDirectionCode(direction);
+  var modecode = getModeCode(mode);
 
-  if (util.isFunction(callback)) {
-    process.nextTick(function() {
-      callback(err, value);
-    });
+  if (dircode < 0 || modecode < 0) {
+    result = gpioctl.ERR_INVALIDPARAM;
   }
   else {
-    return err;
+    result = gpioctl.setPort(portNumber, dircode, mode, callback);
+  }
+  if (result < 0) {
+    err = new GpioError(result, gpioctl.errMessage(result));
+    if (util.isFunction(callback)) {
+      callback(err);
+    }
+    else {
+      throw err;
+    }
+  }
+};
+
+
+GPIO.writePort = function(portNumber, value, callback) {
+  var result = gpioctl.writePort(portNumber, value, callback);
+  if (result < 0) {
+    err = new GpioError(result, gpioctl.errMessage(result));
+    if (util.isFunction(callback)) {
+      callback(err);
```

Comment:
same here. native code does not call `callback` in case of error, just return err code.